sim/sw/include/Broadcast.hpp

```diff
@@ -0,0 +1,152 @@
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _COYOTE_BROADCAST_HPP_
+#define _COYOTE_BROADCAST_HPP_
+
+#include <mutex>
+#include <condition_variable>
+
+/**
+ * Broadcast that has ordering. The way to use this is:
+ * 1. Register yourself as a receiver which returns the last generation
+ * 2. Call receive_any in a loop while incrementing the generation counter until you get the desired 
+ *    output.
+ * 3. Unregister yourself as a receiver
+ * 
+ * The broadcast only continues when it gets acknowledgements from all receivers.
+ */
+template <typename T>
+class Broadcast {
+public:
+    Broadcast() : num_receivers(0), is_broadcast_active(false), curr_generation(0), ack_count(0) {}
+
+    /**
+     * Register a new receiver for broadcasts.
+     * 
+     * @return The current generation that has to be passed ot receive(...)
+     */
+    size_t register_receiver() {
+        std::lock_guard<std::mutex> lock(mtx);
+        num_receivers++;
+        return curr_generation;
+    }
+
+    /**
+     * Unregister a receiver. Wakes up the broadcast if there is an active broadcast and all other 
+     * receivers have acknowledged already.
+     */
+    void unregister_receiver() {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (num_receivers > 0) {
+            num_receivers--;
+
+            if (is_broadcast_active && ack_count >= num_receivers) {
+                ack_cv.notify_one();
+            }
+        }
+    }
+
+    /**
+     * Broadcast data to all receivers. Blocks until all receivers have acknowledged receipt.
+     * 
+     * @param data The data to broadcast
+     */
+    void broadcast(const T &data) {
+        std::unique_lock<std::mutex> lock(mtx);
+        
+        if (num_receivers == 0) {
+            return;
+        }
+
+        // Wait for any previous broadcast to complete
+        ack_cv.wait(lock, [this] { 
+            return !is_broadcast_active; 
+        });
+
+        // Set up new broadcast
+        this->data = data;
+        curr_generation++;
+        ack_count = 0;
+        is_broadcast_active = true;
+
+        // Wake up all waiting receivers
+        broadcast_cv.notify_all();
+
+        // Wait for all acknowledgments
+        ack_cv.wait(lock, [this] {
+            return ack_count >= num_receivers;
```

Comment:
Why >= and not ==?

Can there be some weird corner cases where one receiver sends two acks and some neither, but this condition still ends up satisfied.

Reply:
I think we can leave as is.